### PR TITLE
DOP-1018: Preserve source fileid through includes

### PR DESCRIPTION
**Pull request by i80and**
This changes the AST so the Root node contains a `fileid` field indicating the source of this document. During postprocessing, these values are tracked in a stack, and each SAX handler gets the full fileid stack history for the current page context.

### CODE REVIEW

File: snooty/eventparser.py [hunk 1]
```diff
@@ -30,16 +58,16 @@ def get_event_listeners(self, event: str) -> Set[Callable[..., Any]]:
     def fire(
         self,
         event: str,
-        filename: FileId,
+        fileid: Union[FileId, FileIdStack],
```
Comment (sophstad) [inline snooty/eventparser.py:61]: Do you see any benefit to only using a `FileIdStack` (no union) here? It seems like the page enter/exit events could work with a stack of length 1. But maybe I am missing something!

File: snooty/gizaparser/test_steps.py [hunk 2]
```diff
@@ -9,13 +9,29 @@
...
+    #     category.add(path, text, steps)
+    #     assert len(parse_diagnostics) == 0
+    #     assert len(steps) == 4
+    #     return parse_diagnostics
+
+    # def add_child_file() -> List[Diagnostic]:
+    #     steps, text, parse_diagnostics = category.parse(child_path)
+    #     category.add(child_path, text, steps)
+    #     assert len(parse_diagnostics) == 0
+    #     return parse_diagnostics
```
Comment (sophstad) [inline snooty/gizaparser/test_steps.py:34]: Remove?

File: snooty/n.py [hunk 3]
```diff
@@ -45,6 +47,18 @@
 _T = TypeVar("_T")
 
 
+class FileId(PurePosixPath):
```
Comment (sophstad) [inline snooty/n.py:50]: Nice, this is a better home for this class.

**Review by sophstad — APPROVED** [review]
Nice, this looks good to me! Just left one thought.

In terms of breaking changes, I think we'll just need to ensure that the front end safely ignores the new `root` nodes. Thanks a lot for this Andrew!

File: snooty/gizaparser/test_steps.py (hunk 2 above)
Comment (i80and) [reply snooty/gizaparser/test_steps.py:34]: oops, yup! 😅

File: snooty/eventparser.py (hunk 1 above)
Comment (i80and) [reply snooty/eventparser.py]: No real strong opinion, so I just went ahead and changed the PR to use FileIdStack everywhere!

**Pull request merged**